### PR TITLE
Push packets and their files into a location

### PR DESCRIPTION
**Pull request by richfitz**
This PR adds support for pushing packets into a location. I've made a couple of assumptions here, namely that we'd only ever want to push complete (unpacked) packets and that the remote is definitely running a file store.

There's obviously some work required to implement this for the http driver (and in the server) but this is a good point to discuss the approach I think

This PR builds on a series of mini PRs that add some required functionality:

* #63 
* #65 
* #66 
* #67

Closes #60 (an earlier attempt at this, that used a zip file for transport)

### CODE REVIEW

File: R/location.R [hunk 1]
```diff
@@ -319,6 +319,49 @@ outpack_location_pull_packet <- function(id, location = NULL, recursive = NULL,
...
+##' @export
+outpack_location_push <- function(packet_id, location, root = NULL) {
+  root <- outpack_root_open(root, locate = TRUE)
+  location_id <- location_resolve_valid(location, root,
+                                        include_local = FALSE,
+                                        allow_no_locations = FALSE)
+  plan <- location_build_push_plan(packet_id, location_id, root)
+
+  if (length(plan$files) > 0 || length(plan$packet_id) > 0) {
+    driver <- location_driver(location_id, root)
```
Comment (r-ash) [inline R/location.R:352]: Could get this driver from `location_build_push_plan` if you wanted to avoid building it again

**Review by r-ash — APPROVED** [review]
This looks great to me, one small comment

File: R/location.R (hunk 1 above)
Comment (hillalex) [reply R/location.R:352]: Or create it earlier and pass driver as an argument to `location_build_push_plan`

**Review by hillalex — APPROVED** [review]
This interface looks really clear to me!

File: R/location.R (hunk 1 above)
Comment (richfitz) [reply R/location.R:352]: Only reason for not doing this is to keep symmetry with the pull plan (which might span several locations so does not take a driver)

**Pull request merged**